parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +622,528 @@ protected GroupType build(String name) {
+    }
+
+    public ListBuilder<BaseGroupBuilder<P, Q>> list(Type.Repetition repetition) {
+      return new ListBuilder<BaseGroupBuilder<P, Q>>(this).repetition(repetition);
+    }
+
+    public ListBuilder<BaseGroupBuilder<P, Q>> requiredList() {
+      return list(Type.Repetition.REQUIRED);
+    }
+
+    public ListBuilder<BaseGroupBuilder<P, Q>> optionalList() {
+      return list(Type.Repetition.OPTIONAL);
+    }
+  }
+
+  /**
+   * A builder for {@link GroupType} objects.
+   *
+   * @param <P> The type that this builder will return from
+   *          {@link #named(String)} when the type is built.
+   */
+  public static class GroupBuilder<P> extends BaseGroupBuilder<P, GroupBuilder<P>> {
+
+    private GroupBuilder(P parent) {
+      super(parent);
+    }
+
+    private GroupBuilder(Class<P> returnType) {
+      super(returnType);
+    }
+
+    @Override
+    protected GroupBuilder<P> self() {
+      return this;
+    }
+  }
+
+  public abstract static class BaseMapBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+      Builder<Q, P> {
+    private static final Type STRING_KEY = Types
+        .required(PrimitiveTypeName.BINARY).as(OriginalType.UTF8).named("key");
+
+    public static class MapKeyBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BasePrimitiveBuilder<P,
+        MapKeyBuilder<P, Q>> {
+      private BaseMapBuilder<P, Q> parent = null;
+
+      public MapKeyBuilder(BaseMapBuilder<P, Q> parent, PrimitiveTypeName type) {
+        super(parent.parent, type);
+        this.parent = parent;
+        repetition(Type.Repetition.REQUIRED);
+      }
+
+      public MapKeyBuilder(BaseMapBuilder<P, Q> parent, Class<P> returnType,
+                           PrimitiveTypeName type) {
+        super(returnType, type);
+        this.parent = parent;
+        repetition(Type.Repetition.REQUIRED);
+      }
+
+      public MapValueBuilder<P, Q> value(PrimitiveTypeName type, Type.Repetition repetition) {
+        parent.setKeyType(build("key"));
+        if (parent.parent != null) {
+          return new MapValueBuilder<P, Q>(parent,
+              type).repetition(repetition);
+        } else {
+          return new MapValueBuilder<P, Q>(parent,
+              parent.returnType, type).repetition(repetition);
+        }
+      }
+
+      public MapValueBuilder<P, Q> requiredValue(PrimitiveTypeName type) {
+        return value(type, Type.Repetition.REQUIRED);
+      }
+
+      public MapValueBuilder<P, Q> optionalValue(PrimitiveTypeName type) {
+        return value(type, Type.Repetition.OPTIONAL);
+      }
+
+      public MapGroupValueBuilder<P, Q> groupValue(Type.Repetition repetition) {
+        parent.setKeyType(build("key"));
+        return new MapGroupValueBuilder<P, Q>(parent).repetition(repetition);
+      }
+
+      public MapGroupValueBuilder<P, Q> requiredGroupValue() {
+        return groupValue(Type.Repetition.REQUIRED);
+      }
+
+      public MapGroupValueBuilder<P, Q> optionalGroupValue() {
```

Comment:
I think this should allow both maps and lists to be used as the value, so you'll need requiredListValue, requiredMapValue, optionalListValue, and optionalMapValue.